subsys/net/ip/net_mgmt.c

```diff
@@ -33,98 +33,65 @@ struct mgmt_event_wait {
 	struct net_if *iface;
 };
 
-static K_SEM_DEFINE(network_event, 0, K_SEM_MAX_LIMIT);
-static K_MUTEX_DEFINE(net_mgmt_lock);
+static K_MUTEX_DEFINE(net_mgmt_callback_lock);
+static K_MUTEX_DEFINE(net_mgmt_event_lock);
 
 K_KERNEL_STACK_DEFINE(mgmt_stack, CONFIG_NET_MGMT_EVENT_STACK_SIZE);
 static struct k_thread mgmt_thread_data;
-static struct mgmt_event_entry events[CONFIG_NET_MGMT_EVENT_QUEUE_SIZE];
 static uint32_t global_event_mask;
 static sys_slist_t event_callbacks = SYS_SLIST_STATIC_INIT(&event_callbacks);
-static int16_t in_event;
-static int16_t out_event;
+
+/* event structure used to prevent increasing the stack usage on the caller thread */
+static struct mgmt_event_entry new_event;
+K_MSGQ_DEFINE(event_msgq,
+	sizeof(struct mgmt_event_entry),
+	CONFIG_NET_MGMT_EVENT_QUEUE_SIZE,
+	sizeof(uint32_t));
```

Comment:
```suggestion
K_MSGQ_DEFINE(event_msgq, sizeof(struct mgmt_event_entry),
	      CONFIG_NET_MGMT_EVENT_QUEUE_SIZE,	sizeof(uint32_t));
```

Reply:
Will fix, I was just wondering about https://docs.zephyrproject.org/latest/kernel/services/data_passing/message_queues.html#c.K_MSGQ_DEFINE. It mentions that 'To ensure that each message is similarly aligned to this boundary, q_msg_size must also be a multiple of q_align.' I see no evidence in all other usages that any struct is packed. Is that needed?

Reply:
I wondered if that is of concern in this case, but came to conclusion that it isn't - we align at 4-byte boundary, so the struct size should be a multiple of 4, which should be the case ,as the final element is a pointer/size_t depending on configuration, and the strucutre isn't packet.

Maybe we could just add some static assert to ensure this condition is fulfilled in case someone touches the structure in the future?

Reply:
Added and fixed